src/runner/test.rs

```diff
@@ -233,7 +235,7 @@ fn run_cargo(
                 Err(e.context(FailureReason::ICE).into())
             } else if ran_out_of_space {
                 Err(e.context(FailureReason::NoSpace).into())
-            } else if !deps.is_empty() {
+            } else if let Some(deps) = deps {
```

Comment:
I'm not sure about this change (setting the deps to an empty set), as the rest of Crater is not really built with support for an empty set. For example, I'm pretty sure [this code would break](https://github.com/rust-lang/crater/blob/4759ec53a64139d728afb9f4864bebe227212104/src/report/analyzer.rs#L42-L52), as it would never add the crate to the tree.

I'd be comfortable merging the first two commits, but not the third.

Reply:
Would it make sense to adjust the parsing of Crate from PackageId to be infallible i.e. by having it fallback to an `Unkown(String)` variant containing the raw package id if it can't be parsed otherwise.
Then the third commit wouldn't be needed while still ensuring that the failure reason would be recorded correctly even though it might not be as nicely associated with the crate.

Reply:
Adding a `Crate::Unknown` would then require implementing other functions of `Crate` like `to_rustwide`.

I think it's fine to miscategorize things as unknown if the pkgid changes again, as it's a clear signal something is wrong.

Reply:
Ok, I have removed the third commit and instead added a new commit which adds more package id spec tests and implements more of the package id spec parsing to make those added tests pass